src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,530 @@
+    }
+
+    return InteractionModel::ClusterEntry::Invalid();
+}
+
+/// Load the attribute information into the specified destination
+///
+/// `info` is assumed to be default-constructed/clear (i.e. this sets flags, but does not reset them).
+void LoadAttributeInfo(const ConcreteAttributePath & path, const EmberAfAttributeMetadata & attribute,
+                       InteractionModel::AttributeInfo * info)
+{
+    info->readPrivilege = RequiredPrivilege::ForReadAttribute(path);
+    if (attribute.IsReadOnly())
+    {
+        info->writePrivilege = RequiredPrivilege::ForWriteAttribute(path);
+    }
+
+    if (attribute.attributeType == ZCL_ARRAY_ATTRIBUTE_TYPE)
+    {
+        info->flags.Set(InteractionModel::AttributeQualityFlags::kListAttribute);
+    }
+
+    if (attribute.MustUseTimedWrite())
+    {
+        info->flags.Set(InteractionModel::AttributeQualityFlags::kTimed);
+    }
+
+    // NOTE: we do NOT provide additional info for:
+    //    - IsExternal/IsSingleton/IsAutomaticallyPersisted is not used by IM handling
+    //    - IsSingleton spec defines it for CLUSTERS where as we have it for ATTRIBUTES
+    //    - Several specification flags are not available (reportable, quieter reporting,
+    //      fixed, source attribution)
+
+    // TODO: Set additional flags:
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kFabricScoped)
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kFabricSensitive)
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kChangesOmitted)
+}
+
+InteractionModel::AttributeEntry AttributeEntryFrom(const ConcreteClusterPath & clusterPath,
+                                                    const EmberAfAttributeMetadata & attribute)
+{
+    InteractionModel::AttributeEntry entry;
+
+    entry.path = ConcreteAttributePath(clusterPath.mEndpointId, clusterPath.mClusterId, attribute.attributeId);
+    LoadAttributeInfo(entry.path, attribute, &entry.info);
+
+    return entry;
+}
+
+InteractionModel::CommandEntry CommandEntryFrom(const ConcreteClusterPath & clusterPath, CommandId clusterCommandId)
+{
+    InteractionModel::CommandEntry entry;
+    entry.path                 = ConcreteCommandPath(clusterPath.mEndpointId, clusterPath.mClusterId, clusterCommandId);
+    entry.info.invokePrivilege = RequiredPrivilege::ForInvokeCommand(entry.path);
+
+    if (CommandNeedsTimedInvoke(clusterPath.mClusterId, clusterCommandId))
+    {
+        entry.info.flags.Set(InteractionModel::CommandQualityFlags::kTimed);
+    }
+
+    if (CommandIsFabricScoped(clusterPath.mClusterId, clusterCommandId))
+    {
+        entry.info.flags.Set(InteractionModel::CommandQualityFlags::kFabricScoped);
+    }
+
+    // TODO: Set additional flags:
+    // entry.info.flags.Set(InteractionModel::CommandQualityFlags::kFabricSensitive)
+
+    return entry;
+}
+
+const ConcreteCommandPath kInvalidCommandPath(kInvalidEndpointId, kInvalidClusterId, kInvalidCommandId);
```

Comment:
Can this be constexpr?

Reply:
No:

```
error: the type ‘const chip::app::ConcreteCommandPath’ of ‘constexpr’ variable ‘chip::app::{anonymous}::kInvalidCommandPath’ is not literal
```

Reply:
I wished I had several things constexpr and I cannot (e.g. all the invalid entries/info). I think we should have more of those, but that should probably be a separate pass if we can.